### PR TITLE
support private multizone guest clusters

### PR DESCRIPTION
**Pull request by sjenning**
**What this PR does / why we need it**:
Enables private multizone guest clusters by adding all subnetIDs from all NodePools to the VPC Endpoint.

** Reviewer Notes **
There are four different changes in this PR
- Fix `EnqueueRequestsFromMapFunc` for `HostedCluster` controller to use `ClusterName` from the `NodePoolSpec`, not `ObjectMeta`
- AWS error message can not be trusted to not have per-call unique information e.g. auth failures result in a unique encoded error message from STS.  Replace `awsErr.Message()` with `awsErr.Code()`.
- Add subnet reconciliation logic the `AWSEndpointService` controller
- Add `ec2:ModifyVpcEndpoint` to the CPO Role

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

File: hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go [hunk 1]
```diff
@@ -335,20 +335,6 @@ func reconcileAWSCluster(awsCluster *capiawsv1.AWSCluster, hcluster *hyperv1.Hos
...
 		if hcluster.Spec.Platform.AWS.CloudProviderConfig != nil {
 			awsCluster.Spec.NetworkSpec.VPC.ID = hcluster.Spec.Platform.AWS.CloudProviderConfig.VPC
-
-			// TODO: This https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2728
-			// broke our assumption in CAPA 0.7 for externally managed infrastructure.
-			// This effectively limit our ability to span NodePools across multiple subnets.
-			// In a follow up we need to either enable upstream back to support arbitrary subnets IDs
-			// in the awsMachine CR or possibly expose a slice of available subnets for NodePools in hcluster.Spec.Platform.AWS.
-			if hcluster.Spec.Platform.AWS.CloudProviderConfig.Subnet != nil &&
-				hcluster.Spec.Platform.AWS.CloudProviderConfig.Subnet.ID != nil {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go:345]: will need to drop this commit now https://github.com/openshift/hypershift/pull/950

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 2]
```diff
@@ -181,7 +182,7 @@ func (r *HostedClusterReconciler) SetupWithManager(mgr ctrl.Manager, createOrUpd
 		if !ok {
 			return []reconcile.Request{}
 		}
-		return []reconcile.Request{{NamespacedName: types.NamespacedName{Namespace: nodePool.GetNamespace(), Name: nodePool.ClusterName}}}
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:184]: nice catch, thanks!

File: api/v1alpha1/endpointservice_types.go [hunk 3]
```diff
@@ -26,6 +26,10 @@ const (
 type AWSEndpointServiceSpec struct {
 	// The name of the NLB for which an Endpoint Service should be configured
 	NetworkLoadBalancerName string `json:"networkLoadBalancerName,omitempty"`
+
+	// SubnetIDs is the list of subnet IDs to which guest nodes can attach
```
Comment (enxebre) [inline api/v1alpha1/endpointservice_types.go:30]: this is required now right? If so should drop omitempty tag

**Comment by enxebre** [comment]
Anything we want to add to the docs https://hypershift-docs.netlify.app/how-to/deploy-aws-private-clusters/?
Only one question https://github.com/openshift/hypershift/pull/951#discussion_r799401280
/lgtm
/hold
@sjenning feel free to cancel

**Comment by sjenning** [comment]
Yes, I'll update the docs today with the new user-visible `--aws-zones` flag.  All the rest is hidden away (subnet syncing and infra creation).

I have another PR that I'll address https://github.com/openshift/hypershift/pull/951#discussion_r799401280

/hold cancel

**Pull request merged**